### PR TITLE
RavenDB-7574: WillCauseHardPageFault

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+namespace Sparrow.Platform.Posix
+{
+    public static unsafe class PosixMemoryQueryMethods
+    {
+        public static bool WillCauseHardPageFault(IntPtr addr, long length)
+        {
+            Debug.Assert(addr.ToInt64() % Syscall.PageSize == 0);
+
+            var vecSize = (length + Syscall.PageSize - 1) / Syscall.PageSize;
+            var vec = Marshal.AllocHGlobal((int)vecSize);
```
Comment (ayende) [inline src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs]: You are leaking this pointer.

File: src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                wsInfo[i].VirtualAddress = (byte*)addr.ToPointer() + (i * pagesize);
+
+            fixed (void* pWsInfo = wsInfo)
+            {
+                if (QueryWorkingSetEx(GetCurrentProcess(), (byte*)pWsInfo, (uint)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION) * pages)) == false)
+                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {addr.ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
+
+            }
+
+            for (int i = 0; i < pages; i++)
```
Comment (ayende) [inline src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs]: I think that the logic here needs to be reversed, given the new method name.

File: src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            if (Syscall.mincore(addr.ToPointer(), new IntPtr(length), (char*)vec.ToPointer()) != 0)
+                throw new MemoryInfoException($"Failed to mincore addr: {addr.ToInt64()}, with length: {length}. Last Error = {Marshal.GetLastWin32Error()}");
+
+            for (var i = 0; i < vecSize; i++)
```
Comment (ayende) [inline src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs]: I think that the logic here needs to be reversed, given the new method name.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/LowMemory/MemoryInformation.cs [hunk 4]
```diff
@@ -130,7 +130,7 @@ public static bool IsSwappingOnHddInsteadOfSsd()
             return CheckPageFileOnHdd.WindowsIsSwappingOnHddInsteadOfSsd();
         }
 
-        public static bool WillCauseHardPageFault(IntPtr addr, long length) => PlatformDetails.RunningOnPosix ? PosixMemoryQueryMethods.WillCauseHardPageFault(addr, length) : Win32MemoryQueryMethods.WillCauseHardPageFault(addr, length);
+        public static bool WillCauseHardPageFault(IntPtr addr, long length) => PlatformDetails.RunningOnPosix ? !PosixMemoryQueryMethods.WillCauseHardPageFault(addr, length) : !Win32MemoryQueryMethods.WillCauseHardPageFault(addr, length);
```
Comment (ayende) [inline src/Sparrow/LowMemory/MemoryInformation.cs]: No, this is _REALLY_ confusing. We'll typically look at the impl, not the wrapper

File: src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs [hunk 5]
```diff
@@ -7,23 +7,45 @@ namespace Sparrow.Platform.Posix
...
+        public static bool WillCauseHardPageFault(byte* addr, long length)
         {
-            Debug.Assert(addr.ToInt64() % Syscall.PageSize == 0);
+            Debug.Assert(new IntPtr(addr).ToInt64() % Syscall.PageSize == 0);
 
-            var vecSize = (length + Syscall.PageSize - 1) / Syscall.PageSize;
-            var vec = Marshal.AllocHGlobal((int)vecSize);
-            if (Syscall.mincore(addr.ToPointer(), new IntPtr(length), (char*)vec.ToPointer()) != 0)
-                throw new MemoryInfoException($"Failed to mincore addr: {addr.ToInt64()}, with length: {length}. Last Error = {Marshal.GetLastWin32Error()}");
+            var vecSize = (int)((length + Syscall.PageSize - 1) / Syscall.PageSize);
```
Comment (ayende) [inline src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs]: Overflow?

File: src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs [hunk 6]
```diff
@@ -25,19 +25,19 @@ struct PPSAPI_WORKING_SET_EX_INFORMATION
...
             var wsInfo = new PPSAPI_WORKING_SET_EX_INFORMATION[pages];
             for (var i = 0; i < pages; i++)
-                wsInfo[i].VirtualAddress = (byte*)addr.ToPointer() + (i * pagesize);
+                wsInfo[i].VirtualAddress = addr + (i * pagesize);
 
             fixed (void* pWsInfo = wsInfo)
             {
                 if (QueryWorkingSetEx(GetCurrentProcess(), (byte*)pWsInfo, (uint)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION) * pages)) == false)
-                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {addr.ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
+                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {new IntPtr(addr).ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
```
Comment (ayende) [inline src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs]: The logic here is still wrong.

File: src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs [hunk 7]
```diff
@@ -27,27 +27,38 @@ struct PPSAPI_WORKING_SET_EX_INFORMATION
...
-                wsInfo[i].VirtualAddress = addr + (i * pagesize);
 
-            fixed (void* pWsInfo = wsInfo)
+            PPSAPI_WORKING_SET_EX_INFORMATION* pWsInfo;
+            if (pages > 2)
             {
-                if (QueryWorkingSetEx(GetCurrentProcess(), (byte*)pWsInfo, (uint)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION) * pages)) == false)
-                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {new IntPtr(addr).ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
-
+                var wsInfo = Marshal.AllocHGlobal((int)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION)*pages));
```
Comment (ayende) [inline src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs]: You are leaking this memory!

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs (hunk 7 above)
Comment (aviviadi) [reply src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs]: https://github.com/ravendb/ravendb/pull/3032/commits/e3c3071945eea947886110bdbd7ec3b8f51337eb

**Pull request merged**